`chezmoi chattr +encrypted` with external files cause panic

## Describe the bug

`chezmoi chattr +encrypted` on external files, cause segfault

## To reproduce

```bash
$ chezmoi chattr --debug --verbose +encrypted /external/file
time=1970-01-01T00:00:00.001-00:00 level=INFO msg=persistentPreRunRootE version.version=2.52.4 version.commit="" version.date=2024-10-10T12:20:02Z version.builtBy="" args="[chezmoi chattr --debug --verbose +encrypted /external/file]" goVersion=go1.23.2
time=1970-01-01T00:00:00.001-00:00 level=INFO msg=Stat component=system name=/chezmoi
time=1970-01-01T00:00:00.001-00:00 level=INFO msg=Stat component=system name=/chezmoi/.git
time=1970-01-01T00:00:00.002-00:00 level=INFO msg=ReadFile component=system name=/chezmoi/.chezmoiroot size=5 data="home\n"
time=1970-01-01T00:00:00.002-00:00 level=INFO msg=ReadFile component=system name=/chezmoi/.chezmoiversion size=7 data="2.52.3\n"
time=1970-01-01T00:00:00.002-00:00 level=INFO msg=Stat component=system name=/chezmoi/home
time=1970-01-01T00:00:00.002-00:00 level=INFO msg=Stat component=system name=/chezmoi/home
time=1970-01-01T00:00:00.002-00:00 level=INFO msg=ReadDir component=system name=/chezmoi/home
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x57c5a5cd8740]

goroutine 1 [running]:
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*SourceStateFile).Contents(...)
	/build/chezmoi/src/chezmoi-2.52.4/internal/chezmoi/sourcestateentry.go:164
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).runChattrCmd(0xc0006f4008, 0x57c5a76d40c0?, {0xc00037db40, 0x2, 0x4}, 0xc0000ed860)
	/build/chezmoi/src/chezmoi-2.52.4/internal/cmd/chattrcmd.go:201 +0x620
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).newChattrCmd.(*Config).makeRunEWithSourceState.func2(0xc0006fc308, {0xc00037db40, 0x2, 0x4})
	/build/chezmoi/src/chezmoi-2.52.4/internal/cmd/config.go:1611 +0x85
github.com/spf13/cobra.(*Command).execute(0xc0006fc308, {0xc00037db00, 0x4, 0x4})
	/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985 +0xaaa
github.com/spf13/cobra.(*Command).ExecuteC(0xc0006fa008)
	/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
	/build/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).execute(0xc000931de0?, {0xc0000401f0, 0x5, 0x5})
	/build/chezmoi/src/chezmoi-2.52.4/internal/cmd/config.go:1244 +0x78
github.com/twpayne/chezmoi/v2/internal/cmd.runMain({{0x57c5a61cacd0, 0x6}, {0x0, 0x0}, {0x57c5a61cbd70, 0x14}, {0x0, 0x0}}, {0xc0000401f0, 0x5, ...})
	/build/chezmoi/src/chezmoi-2.52.4/internal/cmd/cmd.go:271 +0x22a
github.com/twpayne/chezmoi/v2/internal/cmd.Main({{0x57c5a61cacd0, 0x6}, {0x0, 0x0}, {0x57c5a61cbd70, 0x14}, {0x0, 0x0}}, {0xc0000401f0, 0x5, ...})
	/build/chezmoi/src/chezmoi-2.52.4/internal/cmd/cmd.go:107 +0xa5
main.main()
	/build/chezmoi/src/chezmoi-2.52.4/main.go:30 +0x107
```

## Expected behavior

Should throw an non-critical error

## Output of command with the `--verbose` flag

See above.

## Output of `chezmoi doctor`

nil

## Additional context

`chezmoi chattr +remove external exact...` will throw error: `chezmoi: rename /chezmoi/folder/create_empty_ /chezmoi/folder/create_: no such file or directory`

--- a/internal/cmd/catcmd.go
+++ b/internal/cmd/catcmd.go
@@ -27,7 +27,7 @@ func (c *Config) newCatCmd() *cobra.Command {
 }
 
 func (c *Config) runCatCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, nil)
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{})
 	if err != nil {
 		return err
 	}

--- a/internal/cmd/chattrcmd.go
+++ b/internal/cmd/chattrcmd.go
@@ -151,7 +151,8 @@ func (c *Config) runChattrCmd(cmd *cobra.Command, args []string, sourceState *ch
 	}
 
 	targetRelPaths, err := c.targetRelPaths(sourceState, args[1:], &targetRelPathsOptions{
-		recursive: c.chattr.recursive,
+		mustNotBeExternal: true,
+		recursive:         c.chattr.recursive,
 	})
 	if err != nil {
 		return err

--- a/internal/cmd/chattrcmd.go
+++ b/internal/cmd/chattrcmd.go
@@ -150,7 +150,7 @@ func (c *Config) runChattrCmd(cmd *cobra.Command, args []string, sourceState *ch
 		return err
 	}
 
-	targetRelPaths, err := c.targetRelPaths(sourceState, args[1:], &targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args[1:], targetRelPathsOptions{
 		mustNotBeExternal: true,
 		recursive:         c.chattr.recursive,
 	})

--- a/internal/cmd/config.go
+++ b/internal/cmd/config.go
@@ -647,7 +647,7 @@ func (c *Config) applyArgs(
 			return err
 		}
 	default:
-		targetRelPaths, err = c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+		targetRelPaths, err = c.targetRelPaths(sourceState, args, targetRelPathsOptions{
 			recursive: options.recursive,
 		})
 		if err != nil {
@@ -2534,7 +2534,7 @@ func (c *Config) setEnvironmentVariables() error {
 // sourceAbsPaths returns the source absolute paths for each target path in
 // args.
 func (c *Config) sourceAbsPaths(sourceState *chezmoi.SourceState, args []string) ([]chezmoi.AbsPath, error) {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
 		mustBeInSourceState: true,
 	})
 	if err != nil {
@@ -2567,7 +2567,7 @@ type targetRelPathsOptions struct {
 func (c *Config) targetRelPaths(
 	sourceState *chezmoi.SourceState,
 	args []string,
-	options *targetRelPathsOptions,
+	options targetRelPathsOptions,
 ) (chezmoi.RelPaths, error) {
 	targetRelPaths := make(chezmoi.RelPaths, 0, len(args))
 	for _, arg := range args {
@@ -2583,24 +2583,22 @@ func (c *Config) targetRelPaths(
 		if sourceStateEntry == nil {
 			return nil, fmt.Errorf("%s: not managed", arg)
 		}
-		if options != nil {
-			if options.mustBeInSourceState {
-				if _, ok := sourceStateEntry.(*chezmoi.SourceStateRemove); ok {
-					return nil, fmt.Errorf("%s: not in source state", arg)
-				}
+		if options.mustBeInSourceState {
+			if _, ok := sourceStateEntry.(*chezmoi.SourceStateRemove); ok {
+				return nil, fmt.Errorf("%s: not in source state", arg)
 			}
-			if options.mustNotBeExternal {
-				targetStateEntry, err := sourceStateEntry.TargetStateEntry(c.destSystem, c.DestDirAbsPath.Join(targetRelPath))
-				if err != nil {
-					return nil, err
-				}
-				if targetStateEntry.SourceAttr().External {
-					return nil, fmt.Errorf("%s: is an external", arg)
-				}
+		}
+		if options.mustNotBeExternal {
+			targetStateEntry, err := sourceStateEntry.TargetStateEntry(c.destSystem, c.DestDirAbsPath.Join(targetRelPath))
+			if err != nil {
+				return nil, err
+			}
+			if targetStateEntry.SourceAttr().External {
+				return nil, fmt.Errorf("%s: is an external", arg)
 			}
 		}
 		targetRelPaths = append(targetRelPaths, targetRelPath)
-		if options != nil && options.recursive {
+		if options.recursive {
 			parentRelPath := targetRelPath
 			// FIXME we should not call s.TargetRelPaths() here - risk of
 			// accidentally quadratic

--- a/internal/cmd/config.go
+++ b/internal/cmd/config.go
@@ -2558,6 +2558,7 @@ func (c *Config) targetRelPath(absPath chezmoi.AbsPath) (chezmoi.RelPath, error)
 
 type targetRelPathsOptions struct {
 	mustBeInSourceState bool
+	mustNotBeExternal   bool
 	recursive           bool
 }
 
@@ -2582,9 +2583,20 @@ func (c *Config) targetRelPaths(
 		if sourceStateEntry == nil {
 			return nil, fmt.Errorf("%s: not managed", arg)
 		}
-		if options != nil && options.mustBeInSourceState {
-			if _, ok := sourceStateEntry.(*chezmoi.SourceStateRemove); ok {
-				return nil, fmt.Errorf("%s: not in source state", arg)
+		if options != nil {
+			if options.mustBeInSourceState {
+				if _, ok := sourceStateEntry.(*chezmoi.SourceStateRemove); ok {
+					return nil, fmt.Errorf("%s: not in source state", arg)
+				}
+			}
+			if options.mustNotBeExternal {
+				targetStateEntry, err := sourceStateEntry.TargetStateEntry(c.destSystem, c.DestDirAbsPath.Join(targetRelPath))
+				if err != nil {
+					return nil, err
+				}
+				if targetStateEntry.SourceAttr().External {
+					return nil, fmt.Errorf("%s: is an external", arg)
+				}
 			}
 		}
 		targetRelPaths = append(targetRelPaths, targetRelPath)

--- a/internal/cmd/destroycmd.go
+++ b/internal/cmd/destroycmd.go
@@ -36,7 +36,7 @@ func (c *Config) newDestroyCmd() *cobra.Command {
 }
 
 func (c *Config) runDestroyCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
 		recursive: c.destroy.recursive,
 	})
 	if err != nil {

--- a/internal/cmd/editcmd.go
+++ b/internal/cmd/editcmd.go
@@ -77,7 +77,7 @@ func (c *Config) runEditCmd(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
 		mustBeInSourceState: true,
 	})
 	if err != nil {

--- a/internal/cmd/forgetcmd.go
+++ b/internal/cmd/forgetcmd.go
@@ -28,7 +28,7 @@ func (c *Config) newForgetCmd() *cobra.Command {
 }
 
 func (c *Config) runForgetCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, nil)
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{})
 	if err != nil {
 		return err
 	}

--- a/internal/cmd/mergecmd.go
+++ b/internal/cmd/mergecmd.go
@@ -37,7 +37,7 @@ func (c *Config) newMergeCmd() *cobra.Command {
 }
 
 func (c *Config) runMergeCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
 		mustBeInSourceState: true,
 		recursive:           true,
 	})

--- a/internal/cmd/readdcmd.go
+++ b/internal/cmd/readdcmd.go
@@ -71,7 +71,7 @@ func (c *Config) runReAddCmd(cmd *cobra.Command, args []string, sourceState *che
 		)
 	} else {
 		var err error
-		targetRelPaths, err = c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+		targetRelPaths, err = c.targetRelPaths(sourceState, args, targetRelPathsOptions{
 			recursive: c.reAdd.recursive,
 		})
 		if err != nil {

--- a/internal/cmd/testdata/scripts/issue4012.txtar
+++ b/internal/cmd/testdata/scripts/issue4012.txtar
@@ -1,0 +1,8 @@
+# test that chezmoi chattr +encrypted prints an error message when run on an external file
+! exec chezmoi chattr +encrypted $HOME${/}.file
+stderr 'is an external'
+
+-- home/user/.local/share/chezmoi/.chezmoiexternal.toml --
+[".file"]
+    type = "file"
+    url = "https://example.com/.file"

